### PR TITLE
EnumJsonRule and tests for enum values in JSON files

### PR DESCRIPTION
**Pull request by UnknownNPC**
PR is related to the issue #357. 
Commit contains `EnumJsonRule` class and `io` tests updates for next fields:

- creatures.json
`tags`
`attackAlgorithm`
- items.json
`tags`
- locations.json
`type`

### CODE REVIEW

File: src/main/java/org/mafagafogigante/dungeon/schema/rules/EnumJsonRule.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import com.eclipsesource.json.JsonValue;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+class EnumJsonRule<T extends Enum<T>> extends StringJsonRule {
+
+  private final List<Enum<T>> enumValues = new ArrayList<>();
```
Comment (mafagafogigante) [inline src/main/java/org/mafagafogigante/dungeon/schema/rules/EnumJsonRule.java]: This is the only issue I see. We are performing linear searches on which we have to call toString() on the values. This approach can lead to quadratic behavior which can be a real issue when we start using the JSON validation code to validate mods as some of these enums can grow quite large with time.

Using a `HashSet<String>` should allow for linear lookups.

**Comment by UnknownNPC** [comment]
@mafagafogigante thank you for review. I supposed that `enum list` requires changes to be more clear, thanks for advice. Unfortunately, `Collections.addAll` wasn't a good way.

**Comment by mafagafogigante** [comment]
The problem wasn't filling the list, that was linear. But searching on it could become dangerously slow. Using a O(1) lookup circumvents this issue. It looks done and ready to me. I will merge both PRs you opened into the development branch later today.

**Comment by mafagafogigante** [comment]
Everything rebased on that long-term PR.

It makes it "UnknownNPC committed with mafagafogigante" because I always rebase your commits on top of mine to make Git history easier for me to manage in the future, but don't worry because you are still the author if you check it using git log, for instance.

**Pull request closed without merging**